src/net_processing.cpp

```diff
@@ -1507,17 +1526,44 @@ void PeerManagerImpl::SendPings()
     for(auto& it : m_peer_map) it.second->m_ping_queued = true;
 }
 
-void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid)
+void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid, std::optional<NodeId> from)
 {
-    m_connman.ForEachNode([&txid, &wtxid](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+    bool flood;
+    if (from) {
+        // Flood those transactions which were received either from pre-reconciliation peers, or
+        // inbound reconciliation but NOT via outbound reconciliation. Flooding then is mainly
+        // used for initial propagation of new transactions across a network of reachable nodes
+        // quickly, while most of the bandwidth is still conserved.
+        if (m_reconciliation.IsPeerRegistered(*from)) {
+            flood = !(*m_reconciliation.IsPeerResponder(*from));
+        } else {
+            flood = true;
+        }
+    }
+    else {
+        // Transactions originated locally should be always announced through reconciliation
+        // for the privacy reasons.
+        //
+        // Consider a non-reachable node: the general policy is *flood transactions came from
+        // inbounds* (see above), so non-reachable nodes would never flood any transactions.
+        // Now, if they flood their local transactions, it will be obvious for their immediate peers
+        // that those transactions belong to those non-reachable nodes.
+        //
+        // On the other hand, if non-reachable nodes reconcile those transactions, they would look
+        // like any other transactions of non-reachable nodes.
```

Comment:
Do we have privacy leaks for the following cases not mentioned in this comment ?

A non-reachable node reconciliate with reconciliation-supporting _flooding_ outbound peers ? As it would be a deviation from the expected announcement policy.

A reachable node reconciliate with reconciliation-supporting _flooding_ inbound/outbound peers ? Likewise, it would be a deviation from the expected announcement policy on those links.

I think those cases are handled well at tx-announcement in SendMessages. Though I believe this comment would be better if it said that reconciliation is _downgraded_ to flooding if this the announcement previously selected for the to-be-announced-to peer.

Reply:
Let's continue this same discussion in one place: https://github.com/bitcoin/bitcoin/pull/21515#discussion_r624952992